### PR TITLE
Fix regression in downloading single files from models artifact store

### LINKED ISSUE #10230: [BUG] Model lineage tracking - missing "MLmodel/registered_model_meta" file
### Issues Policy acknowledgement

- [X] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Other

### Willingness to contribute

No. I cannot contribute a bug fix at this time.

### MLflow version

- Client: 2.8.0
- Tracking server: 2.3.1

### System information

- Ubuntu 20.04 LTS
- Python 3.8

### Describe the problem

It seems that latest release of MLflow breaks compatibility between client and server when it runs on older version of MLflow.

https://github.com/mlflow/mlflow/commit/85037652d6caf4a78d8ac73ca734e433e86b1fd9

### Tracking information

```shell
REPLACE_ME
```

### Code to reproduce issue

```
import mlflow
from mlflow.store.artifact.artifact_repository_registry import get_artifact_repository

# server 2.3.1 + client 2.7.1 -> works
# server 2.3.1 + client 2.8.0 -> throws

mlflow.set_tracking_uri('FILL IN')

model_name = 'FILL IN'
model_version = 'FILL IN'
model_uri = f'models:/{model_name}/{model_version}'

artifact_repository = get_artifact_repository(model_uri)
mlmodel_path = artifact_repository.download_artifacts('MLmodel')
```

### Stack trace

```
Traceback (most recent call last):
  File "/app/miniconda/envs/engine/bin/setup-image", line 33, in <module>
    sys.exit(load_entry_point('engine', 'console_scripts', 'setup-image')())
  File "/app/engine/assembly/scripts.py", line 53, in setup_image
    mlmodel_path = artifact_repository.download_artifacts('MLmodel')
  File "/app/miniconda/envs/engine/lib/python3.8/site-packages/mlflow/store/artifact/models_artifact_repo.py", line 173, in download_artifacts
    self._add_registered_model_meta_file(model_path)
  File "/app/miniconda/envs/engine/lib/python3.8/site-packages/mlflow/store/artifact/models_artifact_repo.py", line 143, in _add_registered_model_meta_file
    write_yaml(
  File "/app/miniconda/envs/engine/lib/python3.8/site-packages/mlflow/utils/file_utils.py", line 250, in write_yaml
    raise e
  File "/app/miniconda/envs/engine/lib/python3.8/site-packages/mlflow/utils/file_utils.py", line 240, in write_yaml
    with codecs.open(yaml_file_name, mode="w", encoding=ENCODING) as yaml_file:
  File "/app/miniconda/envs/engine/lib/python3.8/codecs.py", line 905, in open
    file = builtins.open(filename, mode, buffering)
NotADirectoryError: [Errno 20] Not a directory: '/tmp/tmpv6q_d0ps/MLmodel/registered_model_meta'
```

### Other info / logs

```
REPLACE_ME
```

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [X] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by BenWilson2**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/BenWilson2/mlflow/pull/10362?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10362/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10362
```

</p>
</details>

### Related Issues/PRs

Resolve #10230 

### What changes are proposed in this pull request?

Fixed the regression introduced in #9402 wherein no handling logic is provided for non-directory source file resolution, which would attempt to create a registration metadata file using a destination file as a directory path (which raises an Exception).

### How is this PR tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [X] Manual tests

Validated the basic repro is broken in 2.8.0 and is fixed in this PR (test added to validate as well):

``` python
import mlflow
from mlflow.store.artifact.artifact_repository_registry import get_artifact_repository

mlflow.set_tracking_uri("sqlite:///:memory:")

with mlflow.start_run():
    mlflow.sklearn.log_model("I am a model", "model", registered_model_name="foo")

model_uri = f"models:/foo/1"

artifact_repository = get_artifact_repository(model_uri) 
artifact_repository.download_artifacts("requirements.txt")
```

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [X] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [X] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/store/artifact/test_models_artifact_repo.py [hunk 1]
```diff
@@ -133,13 +135,15 @@ def test_models_artifact_repo_add_registered_model_meta_file():
...
     artifact_location = f"s3://blah_bucket/{artifact_path}"
-    artifact_dst_path = f"{dst_path}/{artifact_path}"
+    artifact_dst_path = f"{dst_path}/{artifact_path}/"
     model_name = "MyModel"
     model_version = "12"
 
     with mock.patch.object(
         MlflowClient, "get_model_version_download_uri", return_value=artifact_location
-    ), mock.patch("mlflow.store.artifact.models_artifact_repo.write_yaml") as write_yaml_mock:
+    ), mock.patch("os.path.isdir", return_value=True), mock.patch(
```
Comment (harupy) [inline tests/store/artifact/test_models_artifact_repo.py:144]: Do we really need to mock `os.path.isdir`?

File: mlflow/store/artifact/models_artifact_repo.py [hunk 2]
```diff
@@ -140,16 +144,22 @@ def list_artifacts(self, path):
...
+                    "model_version": self.model_version,
+                },
+                overwrite=True,
+                ensure_yaml_extension=False,
+            )
+        else:
+            _logger.warning(
+                "Registered Model Metadata file not able to be written. Destination path "
+                f"'{model_path}' is not a directory."
+            )
```
Comment (harupy) [inline mlflow/store/artifact/models_artifact_repo.py:162]: This warning might scare users.

Comment (harupy) [reply mlflow/store/artifact/models_artifact_repo.py:162]: Can we remove the else clause?

File: tests/store/artifact/test_models_artifact_repo.py [hunk 3]
```diff
@@ -155,6 +159,34 @@ def test_models_artifact_repo_add_registered_model_meta_file():
...
+    artifact_location = f"s3://blah_bucket/{artifact_path}"
+    artifact_dst_file_path = f"{dst_path}/{artifact_path}/MLModel.yaml"
+
+    with mock.patch.object(
+        MlflowClient, "get_model_version_download_uri", return_value=artifact_location
+    ), mock.patch(
+        "mlflow.store.artifact.models_artifact_repo.write_yaml"
+    ) as write_yaml_mock, mock.patch(
+        "os.path.isdir", return_value=False
+    ), mock.patch(
```
Comment (harupy) [inline tests/store/artifact/test_models_artifact_repo.py:181]: Same here: https://github.com/mlflow/mlflow/pull/10362/files#r1390275186

File: tests/store/artifact/test_models_artifact_repo.py [hunk 4]
```diff
@@ -155,6 +159,34 @@ def test_models_artifact_repo_add_registered_model_meta_file():
             overwrite=True,
             ensure_yaml_extension=False,
         )
+    # Calling the download_artifacts method on local FileStore will create an ./mlruns directory
+    # which is a test side effect. Clean this up.
+    mlruns_dir = "./mlruns"
+    if os.path.exists(mlruns_dir):
+        shutil.rmtree(mlruns_dir)
```
Comment (harupy) [inline tests/store/artifact/test_models_artifact_repo.py:166]: Instead, can we set the artifact location to a temporary directory? If we hit an error before this block, `./mlruns` won't be cleaned up.

File: mlflow/store/artifact/models_artifact_repo.py [hunk 5]
```diff
@@ -140,16 +144,22 @@ def list_artifacts(self, path):
...
-            model_path,
-            REGISTERED_MODEL_META_FILE_NAME,
-            {
-                "model_name": self.model_name,
-                "model_version": self.model_version,
-            },
-            overwrite=True,
-            ensure_yaml_extension=False,
-        )
+        if os.path.isdir(model_path):
```
Comment (harupy) [inline mlflow/store/artifact/models_artifact_repo.py:147]: Is this file created when we download a subdirectory in the model directory?

```
- model
    - subdir <- this directory
    - MLmodel
    - ...
```

Comment (serena-ruan) [reply mlflow/store/artifact/models_artifact_repo.py:147]: Could we update 'download_artifacts' function to only call this method when 'model_path' is a directory instead?

File: mlflow/store/artifact/models_artifact_repo.py (hunk 2 above)
Comment (BenWilson2) [reply mlflow/store/artifact/models_artifact_repo.py:162]: Agreed. Removed this. It's really only valid for deploying a root model directory and only if deploying to MLserver anyway. I'll add a note in the code about why the file might not be created.

File: mlflow/store/artifact/models_artifact_repo.py (hunk 5 above)
Comment (BenWilson2) [reply mlflow/store/artifact/models_artifact_repo.py:147]: @harupy good call. Adjusted the logic to only create this registered metadata file iff the model_path resolution after download contains the MLmodel file. Add a test to verify this behavior.

@serena-ruan great idea - updated!

File: tests/store/artifact/test_models_artifact_repo.py (hunk 1 above)
Comment (BenWilson2) [reply tests/store/artifact/test_models_artifact_repo.py:144]: refactored tests to use tmp_path location

File: mlflow/store/artifact/models_artifact_repo.py [hunk 6]
```diff
@@ -169,8 +173,14 @@ def download_artifacts(self, artifact_path, dst_path=None):
         :return: Absolute path of the local filesystem location containing the desired artifacts.
         """
 
+        from mlflow.models.model import MLMODEL_FILE_NAME
```
Comment (BenWilson2) [inline mlflow/store/artifact/models_artifact_repo.py:176]: Cannot do a non-local import due to circular reference. We really should migrate constants like these to a separate 'neutral zero-dependency' location, but that is out of scope for this PR.

File: tests/store/artifact/test_models_artifact_repo.py [hunk 7]
```diff
@@ -111,50 +111,99 @@ def test_models_artifact_repo_init_with_stage_uri_and_not_using_databricks_regis
...
+        MlflowClient, "get_model_version_download_uri", return_value=artifact_location
+    ), mock.patch.object(
+        ModelsArtifactRepository, "_add_registered_model_meta_file"
+    ) as add_meta_mock:
+        models_repo = ModelsArtifactRepository(f"models:/{model_name}/{model_version}")
+        models_repo.repo = mock.Mock(**{"download_artifacts.return_value": str(model_dir)})
+
+        models_repo.download_artifacts(artifact_location, str(tmp_path))
+
+        add_meta_mock.assert_called_once()
```
Comment (harupy) [inline tests/store/artifact/test_models_artifact_repo.py:159]: Instead, can we check that the metadata file exists?

File: tests/store/artifact/test_models_artifact_repo.py [hunk 8]
```diff
@@ -111,50 +111,99 @@ def test_models_artifact_repo_init_with_stage_uri_and_not_using_databricks_regis
...
+    artifact_path = "artifact_directory"
     model_name = "MyModel"
     model_version = "12"
+    artifact_location = f"s3://blah_bucket/{artifact_path}"
+
+    # Create a directory without an MLmodel file
+    dummy_dir = tmp_path / artifact_path
+    dummy_dir.mkdir()
+    dummy_file = dummy_dir / "dummy_file.txt"
+    dummy_file.write_text("dummy content")
```
Comment (harupy) [inline tests/store/artifact/test_models_artifact_repo.py:194]:
```suggestion
    dummy_file.touch()
```

We can use `touch` if the content is not used.

**Review by harupy — APPROVED** [review]
LGTM

File: tests/store/artifact/test_models_artifact_repo.py [hunk 9]
```diff
@@ -111,50 +111,104 @@ def test_models_artifact_repo_init_with_stage_uri_and_not_using_databricks_regis
...
+
 
+@pytest.mark.skipif(is_windows(), reason="This test fails on Windows")
+def test_models_artifact_repo_download_with_real_files(tmp_path):
+    # Simulate an artifact repository
+    temp_remote_storage = tmp_path / "remote_storage"
+    model_dir = temp_remote_storage / "model_dir"
+    model_dir.mkdir(parents=True)
+    mlmodel_path = model_dir / "MLmodel"
+    mlmodel_path.write_text("dummy content")
```
Comment (harupy) [inline tests/store/artifact/test_models_artifact_repo.py:142]:
```suggestion
    mlmodel_path.touch()
```

we can use more `touch`.

File: tests/store/artifact/test_models_artifact_repo.py [hunk 10]
```diff
@@ -111,50 +111,104 @@ def test_models_artifact_repo_init_with_stage_uri_and_not_using_databricks_regis
...
+
+def test_models_artifact_repo_does_not_add_meta_for_file(tmp_path):
+    artifact_path = "artifact_file.txt"
+    model_name = "MyModel"
+    model_version = "12"
     artifact_location = f"s3://blah_bucket/{artifact_path}"
-    artifact_dst_path = f"{dst_path}/{artifact_path}"
+
+    dummy_file = tmp_path / artifact_path
+    dummy_file.write_text("dummy content")
```
Comment (harupy) [inline tests/store/artifact/test_models_artifact_repo.py:174]: same here

File: tests/store/artifact/test_models_artifact_repo.py [hunk 11]
```diff
@@ -111,50 +111,104 @@ def test_models_artifact_repo_init_with_stage_uri_and_not_using_databricks_regis
...
+def test_models_artifact_repo_uses_repo_download_artifacts(tmp_path):
     """
-    ``ModelsArtifactRepository`` should delegate `download_artifacts` to its
-    ``self.repo.download_artifacts`` function.
+    `ModelsArtifactRepository` should delegate `download_artifacts` to its
+    `self.repo.download_artifacts` function.
     """
     artifact_location = "s3://blah_bucket/"
+    dummy_file = tmp_path / "dummy_file.txt"
+    dummy_file.write_text("dummy content")
```
Comment (harupy) [inline tests/store/artifact/test_models_artifact_repo.py:121]: same here
